framework/Source/Operations/LookupFilter.swift

```diff
@@ -1,15 +1,39 @@
+import Foundation
+
 public class LookupFilter: BasicOperation {
     public var intensity:Float = 1.0 { didSet { uniformSettings["intensity"] = intensity } }
     public var lookupImage:PictureInput? { // TODO: Check for retain cycles in all cases here
         didSet {
+            inputTextures.removeValue(forKey: 1)
             lookupImage?.addTarget(self, atTargetIndex:1)
-            lookupImage?.processImage(synchronously: true)
         }
     }
-    
+
     public init() {
         super.init(fragmentFunctionName:"lookupFragment", numberOfInputs:2)
         
         ({intensity = 1.0})()
     }
+
+    public override func newTextureAvailable(_ texture: Texture, fromSourceIndex: UInt) {
+        if fromSourceIndex == 1 {
+            inputTextures[fromSourceIndex] = texture
+            return
+        }
+
+        guard let lookupImage, inputTextures[fromSourceIndex] == nil else {
+            super.newTextureAvailable(texture, fromSourceIndex: fromSourceIndex)
+            return
+        }
+
+        // DispatchSemaphore is Sendable and is a kind of async-safe scoped locking
+        // Using NSLock will cause a warning in Xcode and doesn't work as expected at runtime.
+        let lock = DispatchSemaphore(value: 0)
+        Task {
+            await lookupImage.processImage()
+            super.newTextureAvailable(texture, fromSourceIndex: fromSourceIndex)
+            lock.signal()
+        }
+        lock.wait(timeout: .now() + 1)
+    }
```

Comment:
The section is highly duplicated with others. Is it possible to make it a method of a common super class to inherit from?

Reply:
Do you have any naming suggestions for the superclass, like BinarySourceFilter?

It'll be better to reuse these codes in the common cases.
We plan to deprecate using GPUImage3 and migrate to the CBVisualEffectBuiltIn in the near cycle (maybe just after the cycle 5), so I tried to make the least changes for the current code base since all of them will be deprecated soon in the current plan,
and a superclass would be the least change option in my opinion